### ISSUE TITLE
rustup not patchelfing all its downloaded binaries

### ISSUE THREAD

**Issue by kazcw** [NixOS/nixpkgs#47374]
## Issue description

A Rust toolchain downloaded through the `rustup` package installs a `rust-lld` binary that lacks `libstdc++.so.6`. This doesn't seem to be a problem when using `cargo` or `rustc` with default options, but breaks `rustc` when targeting WASM.

### Steps to reproduce

1. Install `rustup`:
`nix-shell -iA rustup`
2. Use `rustup` to install a toolchain, nightly for example:
`rustup toolchain install nightly-x86_64-unknown-linux-gnu`
3. Observe the missing library:
`ldd ~/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/rust-lld`
(Interesting part of output is: `libstdc++.so.6 => not found`)
4. (Optional): add WASM target to rustup, try to cross-compile something, get a loquacious error that ends with `rust-lld: error while loading shared libraries: libstdc++.so.6: cannot open shared object file: No such file or directory`

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.66, NixOS, 18.03.133231.01f5e794913 (Impala)`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.0.4`
 - channels(root): `"nixos-18.03.133231.01f5e794913, unstable-19.03pre152634.218ce4de508"`
 - channels(kaz): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`

**Comment by kazcw** [comment]
I think patchelf needs to be used to modify any new toolchain that gets installed through rustup, but I'm not sure how to go about doing that. I don't think rustup has any hook mechanism. A wrapper script could inspect rustup's subcommands and try to figure out if a new toolchain is being installed and then tweak its files, but that's kind of gross. For now my worst-is-best workaround is setting LD_LIBRARY_PATH in my rust dev shell so I can move on with my WASM project.

**Comment by kazcw** [comment]
#30941 patched rustup to patchelf its downloads; it's patching everything installed to `.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin` to set-interpreter, but not running on the `rustlib` bin directory, and the file in there needs a different patchelf invocation because it links libstdc++.

**Comment by boomshroom** [comment]
Seconded! This affects me both when compiling WASM and RISC-V.

Oddly, it does appear as though rust-lld is being having ld linked to the nix store, but `libstdc++.so.6` is still a problem.

**Comment by NCrashed** [comment]
Also hit the issue with `wasm-pack`.

**Comment by NCrashed** [comment]
Found a workaround, setting `LD_LIBRARY_PATH` that points to `libstdc++.so.6`. Here my `shell.nix` to work with `wasm-pack`:
``` nix
with import <nixpkgs> {};
let src = fetchFromGitHub {
      owner = "mozilla";
      repo = "nixpkgs-mozilla";
      rev = "b9c99d043b1cb55ee8c08265223b7c35d687acb9";
      sha256 = "0akyhdv5p0qiiyp6940k9bvismjqm9f8xhs0gpznjl6509dwgfxl";
   };
in
with import "${src.out}/rust-overlay.nix" pkgs pkgs;
pkgs.stdenv.mkDerivation rec {
  name = "aerospace-env";
  buildInputs = with pkgs; [
    #latest.rustChannels.nightly.rust
    rustup

    # Other packages
    pkgconfig
    openssl
    git
    nodejs-8_x 
  ];
  LD_LIBRARY_PATH="${stdenv.cc.cc.lib}/lib64:$LD_LIBRARY_PATH";
}
```

**Comment by gebner** [comment]
I believe this was fixed in #91327.
